packages/terminal/src/common/terminal-protocol.ts

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { JsonRpcServer } from '@theia/core/lib/common/messaging/proxy-factory';
+
+export const ITerminalServer = Symbol('ITerminalServer');
+
+export const terminalPath = '/services/terminal';
+
+export interface ITerminalServer extends JsonRpcServer<ITerminalClient> {
+    create(options: object): Promise<number>;
```

Comment:
It would be nice to use the interface instead of `object`, e.g. to discover all subinterfaces.

Reply:
Yes that is what I wanted to do but you can't do method overloading in typescript right ?
That's why I used options:object so the method name could be the same even if the real signature is different.
Is there a better way?

Reply:
After reading all the comments below I'll make an interface per Subclass and use the proper interfaces

Reply:
I thought about something like:
```ts
export interface IBaseTerminalOptions {}
export interface IBaseTerminalServer {
    create(options: IBaseTerminalOptions): Promise<number>;
}
export inteface ITerminalOptions extends IBaseTerminalOptions {
    command: string
}
export interface ITerminalServer extends IBaseTerminalServer {
    create(options: ITerminalOptions): Promise<number>;
}
export inteface IShellOptions extends IBaseTerminalOptions {
    shell: string
}
export interface IShellServer extends IBaseTerminalServer {
    create(options: IShellOptions): Promise<number>;
}
```

Reply:
Actually I mean the interface should be OK but not the classes since ShellServer extends TerminalServer

Reply:
Why Shell cannot directly extend Base?

Reply:
I guess I could have a better base class... and use terminal just for create ... so yeah nervermind let me give it another try

Reply:
> Actually I mean the interface should be OK but not the classes since ShellServer extends TerminalServer

but then you can apply the same technique to classes, extract BaseTerminalServer with an abstract create method and provide 2 subclasses: Shell and Terminal

?

Reply:
Yes that's is what I'm doing now, it makes things a bit weird for some stuff maybe.. but I'll let you see when I'm done